### PR TITLE
Q1DWeighted SANS improvement

### PR DESCRIPTION
**Pull request by MathieuTillet**
**Description of work.**

Allow Q1DWeighted to treat monochromatic data using multiple samples in the same workspace.
Using each bin as a different sample, for the same wavelength, it is now possible to treat monochromatic data faster without having to go through an external loop. It is a part of an ongoing effort to make SANS kinetic data treatment more efficient.

**To test:**
- Load `ExternalData/Testing/Data/UnitTest/ILL/D11B/kinetic.nxs`
- Run `Q1DWeighted` with rebinning parameters `0, 0.002, 0.1` and default options.
- Check that there are 3 spectra.
- Check that those 3 spectra have curves with actual values.

WIP for #31943 . 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Algorithms/src/Q1DWeighted.cpp [hunk 1]
```diff
@@ -506,26 +521,70 @@ void Q1DWeighted::finalize(const MatrixWorkspace_const_sptr &inputWS) {
...
+        YOut[iq] += m_intensities[iout][il][iq] / norm;
+        EOut[iq] += m_errors[iout][il][iq] / (norm * norm);
+        normLambda[iq] += 1.;
+      }
+      PARALLEL_END_INTERUPT_REGION
+    }
+    PARALLEL_CHECK_INTERUPT_REGION
+  }
+
+  if (!m_isMonochromatic) {
```
Comment (gvardany) [inline Framework/Algorithms/src/Q1DWeighted.cpp:587]: Is this check still needed?

File: docs/source/algorithms/Q1DWeighted-v1.rst [hunk 2]
```diff
@@ -30,7 +31,7 @@ However, it can also be used directly, provided that the input data is already c
 I/O
 ---
 
-The input to this algorithm must be a histogram with common wavelength bins for each pixel.
+The input to this algorithm must be a histogram with common wavelength bins for each pixel, or a monochromatic workspace with different sample in each bin.
```
Comment (gvardany) [inline docs/source/algorithms/Q1DWeighted-v1.rst:34]:
```suggestion
The input to this algorithm must be a histogram with common wavelength bins for each pixel, or a monochromatic workspace with different sample or kinetic frame in each bin. In this case the workspace must not be in units of wavelength, but it must have the wavelength present in the sample logs.
```

File: docs/source/algorithms/Q1DWeighted-v1.rst [hunk 3]
```diff
@@ -13,14 +13,15 @@ Performs azimuthal averaging for a 2D SANS data set by going through
...
-Note that the algorithm performs averaging in 2 steps.
+Note that in TOF case, the algorithm performs averaging in 2 steps.
 First, for each wavelength slice of the input it performs azimuthal averaging in radial rings.
 This results in a stack of I(Q)s, one for each input wavelength.
 As a second step, all the I(Q)s corresponding to different wavelengths are averaged again.
 Note that for the TOF data, this is mathematically not equivalent to performing averaging with one step; that is averaging everything into a single I(Q) histogram.
 The latter is done in :ref:`Q1D <algm-Q1D>` algorithm.
 See the :ref:`Rebin <algm-Rebin>` documentation for details about choosing the ``OutputBinning`` parameter.
 
+For monochromatic data, the algorithm considers each bin as belonging to a different sample, and averages in distinct bins that are not merged.
```
Comment (gvardany) [inline docs/source/algorithms/Q1DWeighted-v1.rst:24]:
```suggestion
For monochromatic data, the algorithm considers each bin as belonging to a different sample and/or kinetic frame.
```

**Review by gvardany — COMMENTED** [review]
Looks fine, just a couple of minor remarks:

**Comment by gvardany** [comment]
Works alright too, could you just change the test to calculate and assert also on the wedges?

File: Framework/Algorithms/test/Q1DWeightedTest.h [hunk 4]
```diff
@@ -319,6 +319,57 @@ class Q1DWeightedTest : public CxxTest::TestSuite {
...
+    TS_ASSERT_DELTA(result->y(2)[2], 0.1875, 0.0001)
+    TS_ASSERT(result->isCommonBins())
+    TS_ASSERT_EQUALS(result->getMaxNumberBins(), 50)
+
+    TS_ASSERT_THROWS_NOTHING(
+        wedge1 = std::dynamic_pointer_cast<MatrixWorkspace>(AnalysisDataService::Instance().retrieve("out_wedges_1")))
+    TS_ASSERT_THROWS_NOTHING(
+        wedge2 = std::dynamic_pointer_cast<MatrixWorkspace>(AnalysisDataService::Instance().retrieve("out_wedges_2")))
+    TS_ASSERT_EQUALS(wedge1->getMaxNumberBins(), wedge2->getMaxNumberBins())
+    TS_ASSERT_EQUALS(wedge1->getNumberHistograms(), wedge2->getNumberHistograms());
```
Comment (gvardany) [inline Framework/Algorithms/test/Q1DWeightedTest.h:362]: Thanks for extending the test. Could you also assert on `getNumberHistograms` to be 3 and `isCommonBins` also on the wedges? I think the former is the main check to make sure it is not confused with a tof workspace.

**Review by gvardany — APPROVED** [review]

**Review by DavidFair — APPROVED** [review]
LGTM, lets :shipit:

**Pull request merged**